packages/@aws-cdk/aws-s3/lib/rule.ts

```diff
@@ -142,6 +142,38 @@ export interface LifecycleRule {
   readonly objectSizeGreaterThan?: number;
 }
 
+/**
+ * A point in time when an object should transition to a different storage class.
+ */
+export class TransitionTime {
+
+  /**
+   * A moment defined by a specific date.
+   *
+   * The date value must be in ISO 8601 format. The time is always midnight UTC.
+   */
+  public static at(date: Date): TransitionTime {
+    return { transitionDate: date };
+  }
+
+  /**
+   * A moment defined by the number of days after the object creation.
+   */
+  public static after(duration: Duration): TransitionTime {
+    return { transitionAfter: duration };
+  }
+
+  /**
+   * Indicates when objects are transitioned to the specified storage class.
+   */
+  public readonly transitionDate?: Date;
+
+  /**
+   * Indicates the number of days after creation when objects are transitioned to the specified storage class.
+   */
+  public readonly transitionAfter?: Duration;
```

Comment:
Instead of these two, you should probably have a `render(): any` function that returns one of the necessary fields.

Reply:
In the L1 there are two different properties to be set (See my other [comment] for details). So we would need to know which one it is.

[comment]: https://github.com/aws/aws-cdk/pull/23264#issuecomment-1381726278